pool/src/pool/actions.rs

```diff
@@ -177,50 +176,50 @@ pub fn build_actions_from_request(
                     tokens_out = reserve.to_asset_from_b_token(cur_b_tokens);
                 }
                 from_state.remove_collateral(e, &mut reserve, to_burn);
-                actions.add_for_pool_transfer(&reserve.asset, tokens_out);
+                actions.add_for_pool(&request.asset, &request.user, tokens_out);
                 check_health = true;
                 pool.cache_reserve(reserve);
                 e.events().publish(
                     (
                         Symbol::new(e, "withdraw_collateral"),
-                        request.address.clone(),
+                        request.asset.clone(),
                         from.clone(),
                     ),
                     (tokens_out, to_burn),
                 );
             }
             RequestType::Borrow => {
-                let mut reserve = pool.load_reserve(e, &request.address, true);
+                let mut reserve = pool.load_reserve(e, &request.asset, true);
                 let d_tokens_minted = reserve.to_d_token_up(request.amount);
                 from_state.add_liabilities(e, &mut reserve, d_tokens_minted);
                 reserve.require_utilization_below_max(e);
-                actions.add_for_pool_transfer(&reserve.asset, request.amount);
+                actions.add_for_pool(&request.asset, &request.user, request.amount);
                 check_health = true;
                 pool.cache_reserve(reserve);
                 e.events().publish(
                     (
                         Symbol::new(e, "borrow"),
-                        request.address.clone(),
+                        request.asset.clone(),
                         from.clone(),
                     ),
                     (request.amount, d_tokens_minted),
                 );
             }
             RequestType::Repay => {
-                let mut reserve = pool.load_reserve(e, &request.address, true);
+                let mut reserve = pool.load_reserve(e, &request.asset, true);
                 let cur_d_tokens = from_state.get_liabilities(reserve.index);
                 let d_tokens_burnt = reserve.to_d_token_down(request.amount);
-                actions.add_for_spender_transfer(&reserve.asset, request.amount);
+                actions.add_for_spender(&request.asset, &request.user, request.amount);
                 if d_tokens_burnt > cur_d_tokens {
                     let amount_to_refund =
                         request.amount - reserve.to_asset_from_d_token(cur_d_tokens);
                     require_nonnegative(e, &amount_to_refund);
                     from_state.remove_liabilities(e, &mut reserve, cur_d_tokens);
-                    actions.add_for_pool_transfer(&reserve.asset, amount_to_refund);
+                    actions.add_for_pool(&request.asset, &request.user, amount_to_refund);
```

Comment:
This will make it difficult for users to fully repay loans, as users have to sign for "tokens_in", so it needs to be consistent between blocks.

Every block, interest is accrued, and thus the total loan amount being repaid also changes. So, it a user submits a request of "repay 1000 USDC" and their loan is 900, the protocol will calculate a transfer of 900 USDC (100 refund) and request authorization from the user in the transaction.

If they submit the transaction on the next block, their loan is now 900.00001 USDC and the pool will request 900.00001 USDC (99.99999 refund), which would invalidate their signature.

This is less of an issue when #5 is completed, however we still should support singular repayment operations with `transfer`.